src/agent/src/storage/mod.rs

```diff
@@ -380,8 +381,122 @@ fn mount_storage(logger: &Logger, storage: &Storage) -> Result<()> {
         "mount-options" => options.as_str(),
     );
 
+    let confidential = storage.driver_options.contains(&"confidential=true".into());
+    let ephemeral = storage.driver_options.contains(&"ephemeral=true".into());
+
+    let src_path = if confidential && ephemeral {
+        // NOTE: This whole branch will be replaced by the following
+        // upstream script eventually:
+        // https://github.com/confidential-containers/guest-components/blob/main/confidential-data-hub/storage/scripts/luks-encrypt-storage
+
+        // TODO: Ensure that this allows:
+        //  * Multiple containers (same pod) to mount the same ephemeral
+        //    volume.
+        //  * A container to mount the same ephemeral volume mutliple
+        //    times (under different paths).
+        // => No issue in my testing but need to ensure that this is
+        // only called once per pod.
+
+        // Path to the key file that will be passed to the cryptsetup
+        // commands.
+        const KEY_FILE_PATH: &str = "/tmp/ephemeral_encryption_key";
+
+        // Name of the devmapper that will live under /dev/mapper/.
+        let devmapper_device_name = src_path
+            .file_name()
+            .ok_or_else(|| anyhow!("could not get filename from {:?}", src_path))?
+            .to_str()
+            .ok_or_else(|| anyhow!("could not convert {:?} to string", src_path))?;
+
+        // Full path of the devmapper device. This is a virtual device
+        // that appears as cleartext to the end user and maps to
+        // `storage.source`. Under the hood, writes to this device are
+        // encrypted by the kernel and funnelled to the underlying
+        // logical `storage.source` device.
+        let devmapper_device_path = format!("/dev/mapper/{devmapper_device_name}");
+
+        // Generate a random encryption key and write it to a file.
+        let mut urandom = File::open("/dev/urandom")?;
+        let mut key = [0u8; 32];
+        urandom.read_exact(&mut key)?;
+        let mut key_file = File::create(KEY_FILE_PATH)?;
+        key_file.write_all(&key)?;
+
+        // Encrypt the logical `storage.source` device.
+        let output = Command::new("cryptsetup")
+            .args([
+                // TODO: Validate these encryption settings.
+                "luksFormat",
+                "--type",
+                "luks2",
+                &storage.source,
+                "--key-file",
+                KEY_FILE_PATH,
+                "-v",
+                "--batch-mode",
+                "--sector-size",
+                "4096",
+                "--cipher",
+                "aes-xts-plain64",
+                "--pbkdf",
+                "pbkdf2",
+                "--pbkdf-force-iterations",
+                "1000",
+            ])
+            .output()?;
+        if !output.status.success() {
+            info!(logger, "failed to encrypt device";
+                "status" => output.status.code().unwrap_or(-1),
```

Comment:
Please confirm that you tested these error paths, and the log output looks reasonable. I don't remember seeing this info! syntax before...

Reply:
Confirmed. Copied this pattern from the same file.